### PR TITLE
Add implicit padding to softmax when tensor is in tile layout

### LINKED ISSUE #19781: ttnn.softmax low pcc when input is -338953138925153547590470800371487866880.00000
**Describe the bug**
ttnn.softmax is one of the causes for low PCC in CLIP model.

Input spec:
- inputs are all -338953138925153547590470800371487866880.00000
- input shape: [16, 7, 7]
- datatype: ttnn.bfloat16
- layout: ttnn.TILE_LAYOUT
- dim: -1
- numeric_stable: True

PCC is 0.0.
This spec is found in CLIP model and is one of the causes of low accuracy in that model.

**To Reproduce**
Run unit test:
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback
from models.utility_functions import torch_random
from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten


def max_relative_error(expected, calculated):
    assert expected.shape == calculated.shape, "Tensors must have the same shape"
    if expected.dtype == torch.bool:
        expected = expected.to(torch.int)
    if calculated.dtype == torch.bool:
        calculated = calculated.to(torch.int)
    relative_error = torch.abs(expected - calculated) / (torch.abs(expected) + 1e-8)
    mre = torch.max(relative_error).item()
    inds = torch.where(relative_error == mre)
    return mre, inds


def run_softmax_tests(
    input_shape, 
    dtype, 
    dlayout,
    dim,
    numeric_stable,
    device,
):

    x = torch.full(size=input_shape[0], fill_value=-338953138925153547590470800371487866880.00000, dtype=torch.bfloat16)

    try:
        ref_value = aten._softmax.default(x, dim, False)
        
        tt_x = ttnn.from_torch(x, dtype=dtype[0], layout=dlayout[0], device=device)

        tt_result = ttnn.softmax(tt_x, dim, numeric_stable=numeric_stable)
        tt_result = ttnn.to_torch(tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    assert_with_pcc(ref_value, tt_result, 0.999)


test_sweep_args = [
    (
        [[16, 7, 7]],
        [ttnn.bfloat16],
        [ttnn.TILE_LAYOUT],
        -1, 
        True,
    ),
]


@pytest.mark.parametrize(
    "input_shape, dtype, dlayout, dim, numeric_stable",
    (test_sweep_args),
)
def test_softmax(input_shape, dtype, dlayout, dim, numeric_stable, device):
    run_softmax_tests(input_shape, dtype, dlayout, dim, numeric_stable, device)

```

### PR DESCRIPTION
**Pull request by nsorabaTT**
### Ticket
[Addresses 19781](https://github.com/tenstorrent/tt-metal/issues/19781) 

### Problem description
When last dimension is not a multiple of 32 and input tensor is tiled, softmax operates on the padded data. This padded data should be set to -inf for it to not impact softmax values.

### What's changed
1. Adds -inf padding for padded areas before running softmax
2. Adds unit test for coverage.

### Performance Analysis

We consider two options to implement this fix:

1. Add an `implicit_fill` op before calling softmax
2. Use writer kernel to stream _-infinity_ values to the padded areas.

I ran all the softmax unit tests to get some statistics via tracy. Each OP was run twice and data collected from the second run to negate any impact of stale program cache.

From the 330 softmax operations that were run, approach (2) outperformed approach (1) 65% of the time. In the rest off the 35% traces, approach (1) did better marginally. This is evident by looking at relative runtime between (2) to (1). i.e. when we look at how big (or small) Approach (2) is in runtime as compared to Approach(1), we see this:

| Statistic | (Runtime (1) / Runtime (2) (%)     |
|-----------|-----------|
| MIN       | 92.48%    |
| MAX       | 389.99%   |
| MEDIAN    | 100.41%   |
| AVERAGE   | 110.52%   
| TOTAL COUNT of Softmax ops executed | 330 |
| (2) better in | 213 |
| (2) better (%) | 65% |

i.e. even when approach (1) does better, it is only 8% better than approach (2). However, when approach (2) does better, it delivers a huge speed up. In the worst case scenario, the execution time is cut to 1/4th.

To conclude, we go with approach (2) and [create a new work item](https://github.com/tenstorrent/tt-metal/issues/20171) to explore similar approach for other reductions like min/max etc where padding is necessary.

Raw data is [here](https://docs.google.com/spreadsheets/d/18JBgL99okOoqJx3DTg1G5FECPojHaWeE4rtT53Nx35Q/edit?usp=sharing)

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14236083253) CI passing
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes: Added new unit tests

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp [hunk 1]
```diff
@@ -117,8 +135,12 @@ ttnn::Tensor ExecuteScaleCausalMaskHWSoftmaxInPlace::invoke(
     const std::optional<const DeviceComputeKernelConfig> compute_kernel_config,
     const bool numeric_stable) {
     const auto& input_shape = input_tensor.get_logical_shape();
+    bool is_tiled = input_tensor.get_layout() == TILE_LAYOUT;
+    auto padded_input_tensor =
+        is_tiled ? ttnn::fill_implicit_tile_padding(input_tensor, -std::numeric_limits<float>::infinity())
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp:140]: You're calling this code enough times that you should create a function.

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp (hunk 1 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp:140]: I considered it and then opted against because of the overhead for 2 lines. I'll replace with an inlined function as a balance.

**Comment by bbradelTT** [comment]
Why is the following error appearing?
image

**Comment by nsorabaTT** [comment]
> Why is the following error appearing? image

Ack. reproing locally.

**Comment by yugaoTT** [comment]
SMX already has logic in kernel to handle the padding, why is this PR needed?
If that logic is not handling padding well, then we should debug that

File: ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp [hunk 2]
```diff
@@ -7,11 +7,26 @@
...
+ * This function adds -inf value to the padding regions of the input tensor.
+ * It is used to ensure that the softmax operation can be performed correctly
+ * on the input tensor, regardless of its layout.
+ *
+ * @param input_tensor The input tensor to be padded.
+ * @return The padded input tensor.
+ */
+static inline ttnn::Tensor get_padded_input(const ttnn::Tensor& input_tensor) {
+    bool is_tiled = input_tensor.get_layout() == TILE_LAYOUT;
+    return is_tiled ? ttnn::fill_implicit_tile_padding(input_tensor, -std::numeric_limits<float>::infinity())
```
Comment (yugaoTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp:26]: this is gonna hurt perf, we should let SMX itself handle padding (which support should already be there), instead of calling another op

**Review by yugaoTT — CHANGES_REQUESTED** [review]

**Comment by nsorabaTT** [comment]
> > Why is the following error appearing? image
> 
> Ack. reproing locally.

With this change, the `fill_implicit_tile_padding` call is impacting the program hash whenever the size of the tensor is not a multiple of 32. Thus, in all test cases where the width is not a multiple of 32, the program cache count is N instead of 1.
I traced it all the way to confirm that the softmax kernel's hash is not changing. Atul confirmed that `fill_implicit_tile_padding` returns the same tensor with in-place fill, which means hash should not change with that either. I am exploring if there is a way to log the increments to program cache, since that will narrow down the issue further.

File: ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp (hunk 2 above)
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp:26]: We should fix this internal padding and then compare performance of the two.

**Comment by bbradelTT** [comment]
> > > > Why is the following error appearing? image
> > > 
> > > 
> > > Ack. reproing locally.
> > 
> > 
> > With this change, the `fill_implicit_tile_padding` call is impacting the program hash whenever the size of the tensor is not a multiple of 32. Thus, in all test cases where the width is not a multiple of 32, the program cache count is N instead of 1. I traced it all the way to confirm that the softmax kernel's hash is not changing. Atul confirmed that `fill_implicit_tile_padding` returns the same tensor with in-place fill, which means hash should not change with that either. I am exploring if there is a way to log the increments to program cache, since that will narrow down the issue further.
> 
> Confirming that the program cache count will be 2 instead of 1 because of the `fill_implicit_tile_padding` call. The count will be 2 instead of 1 when either of the last 2 dimensions is NOT a multiple of TILE_HEIGHT/TILE_WIDTH (32).
> 
> We have two options:
> 
> 1. Change the program cache limit to 2 only when last 2 dimensions are not a multiple of 32. We have take temporary perf hit to unblock CLIP and other models. Near term, I can add a new issue to fix padding in the softmax kernel itself.
> 2. Abandon this PR, fix padding of the softmax kernel itself. This will take longer than (1) but will mostly have no perf impact.

You need to go with The second option, fix padding in the softmax kernel itself and figure out why it wasn't working.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp [hunk 3]
```diff
@@ -7,8 +7,10 @@
 // #include "debug/dprint.h"
 
 // H-bcast mask
-FORCE_INLINE void generate_bcast_row_mask(const uint32_t cb_id, const uint32_t num_datum_padded, const uint32_t val) {
-    const uint32_t mask_val = val >> 16;
+FORCE_INLINE void generate_bcast_row_mask(const uint32_t cb_id, const uint32_t num_datum_padded) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp:10]: 1. This is now no longer a generic mask.
Would you be able to do
```
FORCE_INLINE void generate_bcast_row_mask(const uint32_t cb_id, const uint32_t num_datum_padded, const uint32_t val = 0xFF80) {
```
?

2. Did this work for any data type before, or was this broken for all types?

3. Is there an online reference that explains that ff80 is -inf? It seems it's easy to make a mistake since the code used ff00 and your comment said fff0 (https://github.com/tenstorrent/tt-metal/issues/19781#issuecomment-2773940991)

4. Please compare performance for this code vs calling implicit fill without calling generate_bcast_row_mask

For this small shape, as well as some larger shapes.

At least measure the DEVICE KERNEL DURATION [ns] from tracy for the two ops vs one op.

Given the cache issues you ran into, I'm not sure if you'l be able to get an end to end time without compilation.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp [hunk 4]
```diff
@@ -285,7 +286,7 @@ tt::tt_metal::operation::ProgramWithCallbacks scale_mask_softmax_multi_core(
             }
 
             SetRuntimeArgs(program, softmax_kernels_id, core, {0, 0, 0, 0, 0, 0});
-            SetRuntimeArgs(program, writer_kernels_id, core, {0, 0, 0, 0, 0, 0, 0xFF00FF00});
```
Comment (yugaoTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp:288]: you can just set the new value here? no need to hardcode in kernel

Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp:288]: Since the pad value does not change, I opted to reduce an argument that we need to assemble in the host code and pass it to the device code.

Comment (yugaoTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/multi_core/softmax_op_multi_core.cpp:288]: I see, yeah I guess either is working.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp (hunk 3 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp:10]: 1. I can. I wanted to make it easy for readers to understand that the pad value is a constant. Hence opted to remove that as an argument.
2. The unit tests break for _all_ data types. The reason we don't see it is because existing mask is _big negative enough_ that you needed to values close to `-infinity` to see this bug.
3. Thanks for pointing out. Fixed the comment in the issue. [Online reference](https://flop.evanau.dev/brainfloat-converter). Please select `bfloat16` and add `FF80` in the hex field. You will see `-infinity` as the value stored. Image below for reference.
4. Pending, will update this comment once I have the results.

image

Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp:10]: 1. 
Actually please define a constexpr variable and pass that in to the function.

If the overhead is much lower than using implicit fill then we will need to put this in a common location and use it for all the ops, where the values might vary, with most likely values being 0, inf, and -inf.

Besides that, the function name is not valid if the value is hardcoded.

3. Why are you saying uint in the comment below when you mean fp16? Also, why does the fp16 value work for fp32? If we change the compute kernel config, will this no longer work?

Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/dataflow/writer_unary_interleaved_start_id_blocked_sm.cpp]: (1) Incorporated suggestions in latest commit.
(2) Addressed above.
(3) Clarified for in latest commit to say 0xFF80 is the uint16 representation of bfloat's -inf.
(4) Added detailed performance analysis in the PR description. Raw data linked.

File: ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp (hunk 2 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/softmax.cpp]: Added perf numbers and updated appraoch to fix padding in SMX

**Review by bbradelTT — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]
